store/tikv/client.go

```diff
@@ -15,135 +15,178 @@
 	readTimeoutMedium = 60 * time.Second  // For requests that may need scan region.
 	readTimeoutLong   = 150 * time.Second // For requests that may need scan region multiple times.
+
+	rpcLabelKV  = "kv"
+	rpcLabelCop = "cop"
 )
 
+// Client is a client that sends RPC.
+// It should not be used after calling Close().
+type Client interface {
+	// Close should release all data.
+	Close() error
+	// SendReq sends Request.
+	SendReq(ctx goctx.Context, addr string, req *tikvrpc.Request) (*tikvrpc.Response, error)
+}
+
+// TODO: Add flow control between RPC clients in TiDB ond RPC servers in TiKV.
+// Since we use shared client connection to communicate to the same TiKV, it's possible
+// that there are too many concurrent requests which overload the service of TiKV.
 type rpcClient struct {
-	msgID uint64
-	p     *Pools
+	// TODO: Refactor ConnPools
+	p *Pools
 }
 
 func newRPCClient() *rpcClient {
 	return &rpcClient{
-		msgID: 0,
 		p: NewPools(maxConnection, func(addr string) (*Conn, error) {
 			return NewConnection(addr, dialTimeout)
 		}),
 	}
 }
 
-// SendCopReq sends a Request to co-processor and receives Response.
-func (c *rpcClient) SendCopReq(ctx goctx.Context, addr string, req *coprocessor.Request, timeout time.Duration) (*coprocessor.Response, error) {
+// SendReq sends a Request to server and receives Response.
+func (c *rpcClient) SendReq(ctx goctx.Context, addr string, req *tikvrpc.Request) (*tikvrpc.Response, error) {
 	start := time.Now()
-	defer func() { sendReqHistogram.WithLabelValues("cop").Observe(time.Since(start).Seconds()) }()
-
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	default:
+	var label = rpcLabelKV
+	if req.Type == tikvrpc.CmdCop {
+		label = rpcLabelCop
 	}
+	defer func() { sendReqHistogram.WithLabelValues(label).Observe(time.Since(start).Seconds()) }()
 
 	conn, err := c.p.GetConn(addr)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	defer c.p.PutConn(conn)
-	msg := msgpb.Message{
-		MsgType: msgpb.MessageType_CopReq,
-		CopReq:  req,
-	}
-	err = c.doSend(conn, &msg, writeTimeout, timeout)
+	resp, err := c.callRPC(ctx, conn, req)
 	if err != nil {
 		conn.Close()
 		return nil, errors.Trace(err)
 	}
-	if msg.GetMsgType() != msgpb.MessageType_CopResp || msg.GetCopResp() == nil {
-		conn.Close()
-		return nil, errors.Trace(errInvalidResponse)
-	}
-	return msg.GetCopResp(), nil
+	return resp, nil
 }
 
-// SendKVReq sends a Request to kv server and receives Response.
-func (c *rpcClient) SendKVReq(ctx goctx.Context, addr string, req *kvrpcpb.Request, timeout time.Duration) (*kvrpcpb.Response, error) {
-	start := time.Now()
-	defer func() { sendReqHistogram.WithLabelValues("kv").Observe(time.Since(start).Seconds()) }()
-
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
+func (c *rpcClient) callRPC(ctx goctx.Context, conn *Conn, req *tikvrpc.Request) (*tikvrpc.Response, error) {
+	resp := &tikvrpc.Response{}
+	resp.Type = req.Type
```

Comment:
There usually two styles to write the code to construct a struct with public fields in golang:
1. If all fields needed is ready, use this form
```
s = &structS {
field1: v1,
field2: v2,
}
```
2. Not all fields needed are ready right at place, set the fields one by one
```
s = new(structS)
// prepare v1 for field1
...
s.field1 = v1
// prepare v2 for field2
...
s.field2 = v2
```
Here the style 2 is taken constantly since the concrete response struct depends on request type.